source/common/json/json_internal.cc

```diff
@@ -787,7 +787,9 @@ ObjectSharedPtr Factory::loadFromProtobufStruct(const ProtobufWkt::Struct& proto
 
 std::string Factory::serialize(absl::string_view str) {
   nlohmann::json j(str);
-  return j.dump();
+  // Call with defaults except in the case of UTF-8 errors which we may replace
+  // invalid UTF-8 characters instead of throwing an exception.
+  return j.dump(-1, ' ', false, nlohmann::json::error_handler_t::replace);
```

Comment:
actually can you leave the FFFD replacement in here, and add an debug-only check using `#ifndef NDEBUG` to help convince ourselves that, in unit tests at least, we never get the replacement, but always avoid that by using the octal escapes?

Reply:
Okay, if the coverage test say yes.

Reply:
Done and hope CI happy.

Reply:
Oh sorry, can we have the same call to j.dump with 4 args before the ifdef, and just do the scan to make sure the substitution did not happen under ifndef, e.g.
```
const std::string ret = j.dump(-1, ' ', false, nlohmann::json::error_handler_t::replace);
ASSERT(ret.find(UnexpectedChar) == std::string::npos);
```
return ret;

Actually we don't need the ifnfdef; the ASSERT does all we need.

IMO it's a bit of an anti-patern to have different code computing the answer for debug vs release, whereas it's OK to call extra code in debug mode to validate the result.

Reply:
This looks better.